### ISSUE TITLE
YOLOV8 - COCO Pretained on only Boat Class

### ISSUE THREAD

**Issue by adriengoleb** [ultralytics/ultralytics#25583]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello,

I would like to pre-train my YOLOv8 model on only the "Boat" class from COCO—that is, only on the "BOAT" labels.
I then want to fine-tune my model on a specific dataset that itself contains only the "BOAT" class.

How can I write a script to meet this need?

Here is my script for now, with the parameter `classes = [8]` to select only the `BOAT` labels : 

```
from ultralytics import YOLO
import os


# Initialize a model from scratch
model = YOLO("P:/AGO/Developpement/Models/YOLO_train/yolov8n.yaml")

# Dossier des runsa
save_dir = "P:/AGO/Developpement/Models/YOLO_train/runs/detect"
os.makedirs(save_dir, exist_ok=True)

# Use the model
if __name__ == '__main__':    
    results = model.train(data='N:/IA/COCO/split_coco/coco_boat.yaml', 
    batch=4, 
    device='cuda',
    epochs= 60,
    patience=15, 
    imgsz=640, 
    classes=[8],

    save_period= 5,
    workers=1, 
    save_dir=save_dir, 
    project = save_dir)

```

And my YAML file: 

```

# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
#path: N:/IA/SMD_0/SMD_IR_v0/SMD_IR/nir2nir  # dataset root dir

path: N:/IA/COCO/split_coco
train: images/train   # train images (relative to 'path') 4 images
val: images/val  # val images (relative to 'path') 4 images
test:  # test images (optional)


# Nombre de classes
nc: 1

# Classes (80 COCO classes)
names:
  0: Boat

```

MAKING SURE TO NOTE that I only have one class!

Is this the right approach?
By specifying both files, won't I run into a problem with class numbering, as I've seen happen before?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @adriengoleb, thank you for your interest in Ultralytics 🚀! This is an automated response, and an Ultralytics engineer will also assist you soon.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for new users, where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and answers to many common questions.

For your COCO-to-single-class training workflow, please include any relevant dataset structure, label examples, training logs, and the exact package and model versions so the team can verify the class mapping and fine-tuning setup.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
YOLOv8 is already pretrained on COCO. You can use it directly for fine-tuning. If you use latest Ultralytics, and make sure the class name in your dataset is exactly "boat", then it will automatically transfer the boat class weights during fine-tuning. You need to make sure your dataset exactly has the class name as "boat", not anything else.

If you want to pretrain again, you also need to set `single_cls=True`. `classes` only filters the classes from dataset, but the model still has 80 classes because the original dataset has 80 classes.

**Comment by adriengoleb** [comment]
Thank you very much for your feedback,

Typically, when I'm fine-tuning, I specify my YAML file as follows: 


```
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
#path: N:/IA/SMD_0/SMD_IR_v0/SMD_IR/nir2nir  # dataset root dir

path: C:/Temp/split_smd_mayottev2_to_mayotte2025v2_eval_one_class
train: images/train  # train images (relative to 'path') 4 images
val: images/val  # val images (relative to 'path') 4 images
test:  # test images (optional)

# Classes (80 COCO classes)
names:
  0: boat

```
Is there anything else I need to do in the train.py ?

Before, I had written "Boat" instead.
Does that mean Yolo has random weights on those heads since the class wasn't selected?

My ultralytics version is 8.3.70

**Comment by Y-T-G** [comment]
Your version of Ultralytics is old. It will randomly initialize the head. You need latest Ultralytics for weight transfer feature. Although, it still wouldn't also work because you're using n size. It works for s and larger sizes.

If you pretrain with `single_cls=True` on COCO boat class, then it will not randomly initialize weights when fine-tuning because the model has same number of classes. It reinitializes if the number of classes does not match.

**Comment by adriengoleb** [comment]
Thanks, I understand it much better now.

I now have the following .py file:

```
from ultralytics import YOLO
import os


# Initialize a model from scratch
model = YOLO("P:/AGO/Developpement/Models/YOLO_train/yolov8n.yaml")

# Dossier des runsa
save_dir = "P:/AGO/Developpement/Models/YOLO_train/runs/detect"
os.makedirs(save_dir, exist_ok=True)

# Use the model
if __name__ == '__main__':    
    results = model.train(data='N:/IA/COCO/split_coco/coco_boat.yaml', 
    single_cls=True,
    
    batch=4, 
    device='cuda',
    epochs= 60,
    patience=15, 
    imgsz=640, 
    classes=[8],

    save_period= 5,
    workers=1, 
    save_dir=save_dir, 
    project = save_dir)


```

But what about the **YAML file**? Should I specify only the class 8 boat, or all ?

**Comment by Y-T-G** [comment]
For your fine-tuning dataset, you will just have one class in YAML. It should be class 0 as boat.

**Comment by adriengoleb** [comment]
Yes but I was originally talking about the YAML and Python files for pre-training

**Comment by Y-T-G** [comment]
If you're using COCO dataset for pretraining, then you need to use the COCO YAML as is.

COCO YAML is provided by Ultralytics. You just need to use data=coco.yaml and Ultralytics auto downloads the dataset and runs training

**Comment by adriengoleb** [comment]
Thanks again for your feedback.
I understand the distinction you made.

My last question is about pre-training on Coco Boat, and more specifically, the choice between starting from scratch with the yolov8.yaml file or starting from the yolov8.pt file.


If I go with the latter, I’ll get a double training to the Boat class (one with the pretrain with global coco and the second with the coco boat).
But if I go with YAML, I’ll certainly get some preliminary practice with Coco Boat, but I won’t learn everything about the outlines and colors of objects using Coco.

So I’m facing a dilemma, and I don’t know which option to choose.

**Comment by Y-T-G** [comment]
I think you should use `.pt`. Starting from scratch can increase false positives

**Comment by adriengoleb** [comment]
It works, but when doing this, 

By specifying `classes=[8]` in `model.train()`, YOLO performs on-the-fly filtering:

  It loads ALL 118,287 images from COCO one by one.

  For images without boats, it removes the bounding boxes for other classes and treats them as background images.

So I’m still training the network on 115,000 images that don’t contain boats, which leads to extreme slowness and the risk of background bias.

**Comment by Y-T-G** [comment]
You will have to manually remove the empty images if you want to exclude them

**Comment by glenn-jocher** [comment]
Closing this answered configuration question. Current [label filtering](https://github.com/ultralytics/ultralytics/blob/5eda95e839a9601fc479f1b5e042d48d7e77fa13/ultralytics/data/base.py) keeps the image list: `classes=[8]` filters annotations, and `single_cls=True` remaps the retained class. It deliberately does not delete negative/background images. Use the original COCO YAML for that pretraining pass; your separately relabeled one-class dataset should use `0: boat` for fine-tuning. If you want a boat-containing subset, explicitly build its image list, as [Y-T-G explained](https://github.com/ultralytics/ultralytics/issues/25583#issuecomment-5189885200). Fewer training images reduces work but is not automatically a better dataset or accuracy outcome. No automatic image-pruning change is justified by this report.
